### DISCUSSION TITLE
Including Android app with the integration

### DISCUSSION THREAD

**Discussion by matejdro** [home-assistant/architecture#581] [General] [answered]
Hello everyone.

This is not so much architecture question, as it is philosophy and the way of doing things the-ha-way.

I’m thinking of developing companion app for Android TV, that would bring local push notifications to the androidtv (currently, androidtv integration constantly polls, which is wasteful and slow. App would be running directly on the Android TV, detect system events and transmit them to the HA directly in real time).

I’ve talked to the maintainer of the androidtv integration and he likes the idea. But currently I’m thinking on what would be the best way to get this app to the users of the integration.

My idea currently is that, since androidtv integration already uses ADB bridge that can install other apps, integration would include the companion app in its files and it would automatically install it to the Android TV and configure it (pass HA token etc.) whenever it connects to the Android TV.

Would this approach (including binary of the otherwise open-source app and automatically installing it to the user’s device) be okay with the Home Assistant? If not, what alternative do you propose? Could we upload the APK in some trusted HA repo and instruct user to download and install it manually?

Thanks.

**Comment by Drafteed** [comment]
Very cool idea! Many times I thought about a companion application for Android TV when using slow ADB integration, but I'm not an android developer :)

> My idea currently is that, since androidtv integration already uses ADB bridge that can install other apps, integration would include the companion app in its files and it would automatically install it to the Android TV and configure it (pass HA token etc.) whenever it connects to the Android TV.

I think that it is better not to use the ADB integration at all and place the application separately so that users themselves download it directly or in the Google Play, as is done with a mobile companion. For new integration users, this will only be an additional step (enable dev mode, turn on ADB Debugging in settings, find IP etc..). Also, on some android devices, ADB debugging is automatically disabled after reboot and I have one such device.

**Reply by matejdro** [reply]
Integration already needs ADB to operate and it will need it even with the companion app. Most HA actions (such as turning the device on) cannot be done by an app and still need ADB. App would only be used to speed up Android TV -> HA data updates. So users will still need to perform additional step (ADB Debugging, find IP etc.).

Play Store sounds like a good solution, but full disclosure, while I am prepared to develop this app and support it via APK, I am not prepated to create a Play Store listing and maintain it (and deal with all mandatory questionnaires, updates and other crap that Google throws onto Play Store developers). Plus, some android TV devices (such as Amazon's Fire Stick) do not have Play Store.

**Reply by Drafteed** [reply]
Sorry, how many actions require ADB to implement full functionality? Spotify can wake up all of my devices from sleep and I don't know how he does it :)

**Reply by matejdro** [reply]
Going through the service list for androidtv integration:

* adb command is off the table (since it sells raw adb commands of course). Out of keys, only power on is possible (as you mentioned with Spotify. I misspoke before.), but not power off or simulating other keys like back or up/down.
* androidtv.learn_sendevent will not work
* download/upload services could work with app to some extend, but ADB has permission to more files (especially on Android 10+)
* changing source (e.g. opening new app) would work, but it would be a bit of a hassle for the user (it needs special permissions that are buried in Android settings screen and I'm not even sure if all Android TV devices have those screens)

**Reply by Drafteed** [reply]
> adb command is off the table (since it sells raw adb commands of course). Out of keys, only power on is possible (as you mentioned with Spotify. I misspoke before.), but not power off or simulating other keys like back or up/down.

In the same time app like Button Mapper can power off devices, open apps, exec intents and simulating all keys without root.

**Reply by matejdro** [reply]
Then I stand corrected. It's possible that Button Mapper uses some method that is unknown to me.

**Comment by rytilahti** [comment]
I don't really understand why you want to create a separate app for this and not simply extend the already existing android companion app to support android tv devices (if that doesn't already work as expected). Mind elaborating why you want to do that?

**Reply by matejdro** [reply]
I feel like projects want to achieve very different purpose.

Android app:

* Gives user full access to the Lovelace (which requires touch / mouse that many Android TV devices do not have)
* Requires user to login with his/hers full HA login (which is cumbersome on many Android TV devices where only input is navigating through the keyboard via D-pad on the remote)
* Has to balance sending updates to HA with battery consumption on the phone (and thus updates may be delayed)

Android TV helper app would:

* Not have any UI at all 
* Ideally login automagically via integration (integration would generate token and send it via ADB to automate setup)
* Not care much about battery life, since Android TV devices are usually plugged into the wall, and thus can provide instant media playback state updates (not sure if regular Android app even has media player component)

Of course the two could be merged together, but I feel it would involve way more work than just creating this small companion separately.

**Reply by balloob** [reply]
I agree with @rytilahti that this should be added to the official HA app. That way it can use the existing play store listing. It doesn't have to offer the full capabilities that the app offers on phones, it can just offer event/state forwarding functionality with a basic UI to allow pairing with HA.

**Reply by matejdro** [reply]
I took a look at Android app code. Currently it works by sending sensor data approximately every 15 minutes. This is understandable for phones, since you need to save battery life, but not useful on this project where whole goal is to have instant updates (as opposed to polling). So for this project, Android app would have to be modified with:

* The entirely new login flow that supports remote control input (and preferably some sort of PIN system since entering your password via remote is a pain)
* Entirely new data sending system that sends data in real time (as opposed to in intervals to save battery)
* Updates to the `mobile_app` integration that would create its own media player component for Android devices (does it even make sense to do that, considering that this is not really mobile? This would need to be rethought)

Modifying existing app with that functionality will require way more work than creating simple companion app without UI that communicates directly with the androidtv integration.

**Reply by balloob** [reply]
That's the phone service. You could create a new service just for Android TV. 

I just don't think that it makes sense to have two official Home Assistant Android apps.

**Reply by Drafteed** [reply]
@balloob I find it completely normal practice to use a separate Android TV / Google TV app. Everyone does this: YouTube, Twitch, Spotify, Netflix, Prime Video, etc .. Separate applications for Android TV are primarily made for a remote control. In addition, in most cases they display a different UI designed for TV.

**Reply by elupus** [reply]
Actually the official google play store recommendation is to provide only one app that works for both cases. It's fully possible for one app to have different UIs.

**Reply by matejdro** [reply]
It makes sense to have one app when that app has the same purpose on all platforms. But this proposed app's goals are way different than official HA app.

**Reply by balloob** [reply]
Right now. The future can bring a remote friendly ui. 

If google suggests a single app, that settles it.

**Reply by dshokouhi** [reply]
> I took a look at Android app code. Currently it works by sending sensor data approximately every 15 minutes. This is understandable for phones, since you need to save battery life, but not useful on this project where whole goal is to have instant updates (as opposed to polling)

Sorry I am responding so late. Actually this is technically not completely true. While our sensor worker is limited to every 15 minutes we have quite a bit of sensors that updates with the intent like Bluetooth, Wifi, Screen etc.. so if a sensor supports push updates, we will indeed push the updates instantly. Sensors that do not have a corresponding intent will update on the schedule as those need to be polled. We also offer faster polling up to 1 minute if a user desires too.

**Comment by elupus** [comment]
IMHO adding what is needed to standard app would be smaller. It already support a lot of sensors that could be useful. It already have most login infrastructure. 

It is already published to play store.

**Reply by matejdro** [reply]
Most of the sensors would not really be useful (they do not apply to the stationary always-plugged device like Android TV). And existing login infrastructure is not really that useful since we would have to recreate login UI from scratch just for Android TV.

**Reply by dshokouhi** [reply]
we actually have a workaround implemented for users to be able to login to the app using the native UI, so it is doable in teh current state of the app to login and gain a token etc...

**Comment by JeffLIrion** [comment]
Just to be clear, @matejdro is proposing a simple, lightweight app that would report state updates to the HA Android TV integration. That's the full scope of what the app would do, as I understand it. The user would need to install the app on their Android TV device in order to enable these local push updates for the HA integration. 

This is similar to how the [Notifications for Android TV / FireTV](https://www.home-assistant.io/integrations/nfandroidtv/) integration requires the [Notifications for Android TV](https://play.google.com/store/apps/details?id=de.cyberdream.androidtv.notifications.google) app. However, @matejdro doesn't want to publish the app to the Google Play store, and he's asking about other ways that HA users could install the app. As he said, it should be possible for the Android TV integration to "automagically" install the app using ADB. The question is, would it be acceptable for HA to do so?

An alternative idea that I just thought of is, when setting up the integration, check if the app is installed (and maybe check its version, too). If the app is not installed, use a persistent notification to provide the user with the `androidtv.adb_command` service call that would install/upgrade the app on the device. 

While it may seem like the obvious solution is to add this functionality to the HA Android companion app, I think the problem is that there's a lot of work needed to make that app run well on Android TV devices, and that may not be something that the app developers want to undertake.

**Reply by tetienne** [reply]
I totally understand the need. But how to be sure the app push through adb is well the correct one and not a corrupted application? 
I don't know if it is possible but we can also display a pin code (like plex does) to trust the source.

**Comment by JeffLIrion** [comment]
I worry that the high barrier to entry has deterred a contribution that would be greatly appreciated by HA users. It sounds like @matejdro has something good, but he doesn't want to take on the tasks of getting an app into the Google Play store or integrating this functionality into the HA Android app. 

Something is better than nothing, so how about this. @matejdro distribute the app however you want. If HA changes are needed to make it work, create a custom component. If this gets some traction, then maybe a developer will help to get it officially integrated into HA.

**Reply by matejdro** [reply]
I will likely create an app in my github repo and then just publish it via Github as an unofficial addon.

**Reply by matejdro** [reply]
I've released the app here: https://github.com/matejdro/AndroidTVRefresher

**Comment by matejdro** [comment, marked as answer]
I've released the app here: https://github.com/matejdro/AndroidTVRefresher
